src/ir/auth_logic/ast_construction_test.cc

```diff
@@ -0,0 +1,90 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "src/common/testing/gtest.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+
+using testing::ElementsAre;
+using testing::Property;
+using testing::ResultOf;
+
+namespace raksha::ir::auth_logic {
+
+std::string GetQueryPrincipalName(const Query &query) {
+  return query.principal().name();
+}
+std::string GetQueryFactName(const Query &query) {
+  return std::get<datalog::Predicate>(
+             std::get<BaseFact>(query.fact().GetValue()).GetValue())
+      .name();
+}
+datalog::Predicate GetQueryFactArgs(const Query &query) {
+  return std::get<datalog::Predicate>(
+      std::get<BaseFact>(query.fact().GetValue()).GetValue());
+}
+
+struct QueryTestData {
+  std::string input_text;
+  std::string query_name;
+  std::string principal_name;
+  std::string fact_name;
+  datalog::Predicate predicate;
+};
+
+class AstConstructionTest : public testing::TestWithParam<QueryTestData> {};
+
+TEST_P(AstConstructionTest, SimpleTestWithQueryProgram) {
+  const auto &[input_text, query_name, principal_name, fact_name, predicate] =
+      GetParam();
+  Program prog_out = ParseProgram(input_text);
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre(Property(&Query::name, query_name)));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryPrincipalName, principal_name))));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryFactName, fact_name))));
+  EXPECT_THAT(prog_out.queries(),
+              ElementsAre((ResultOf(GetQueryFactArgs, predicate))));
+}
+
+INSTANTIATE_TEST_SUITE_P(
+    AstConstructionTest, AstConstructionTest,
+    testing::Values(
+        QueryTestData(
+            {R"(may_notifierA = query "UserA" says mayA("NotifierA"))",
+             "may_notifierA", R"("UserA")", "mayA",
+             datalog::Predicate("mayA", {R"("NotifierA")"},
+                                datalog::Sign::kPositive)}),
```

Comment:
I want to preface this by saying this is entirely optional.

One thing I often like to do with tests like this, wherein the test is ensuring that some string in a parsed language is taken apart into its components correctly, is to define a format string and build the test data from the component pieces using that format string. For instance, in this case, the format string seems to be of the form:

```C++
R"(%s = query %s says %s%s)"
```

By plugging your expected results into this format string, you eliminate repeating them and possibly a source of bugs.

Reply:
+1 to use this approach wherever possible. However, in these tests, I don't know if there is enough repetition of the same pattern to do it.

Reply:
I am going to take a note of this and use this next time!